TildeSql/Internal/QueryEngine.cs

```diff
@@ -203,51 +199,81 @@ public async IAsyncEnumerable<T> GetResult<T>(IQuery query)
-                    var cacheExecutionResult = await entry.Value.ExecuteAsync(queriesStillToExecute, cancellationToken);
-                    foreach (var executedQuery in cacheExecutionResult.ExecutedQueries) {
-                        this.cacheExecutorQueries[entry.Index].Add(executedQuery);
-                    }
+            if (queriesStillToExecute.Any() && this.cacheExecutor != null) {
+                await ExecuteAgainstCacheAsync().ToArrayAsync(cancellationToken);
+            }
 
-                    foreach (var (original, executed, remaining) in cacheExecutionResult.PartiallyExecutedQueries) {
-                        if (this.queryForwardMap.ContainsKey(original)) {
-                            this.queryForwardMap.Add(original, [executed, remaining]);
-                        }
-                        else {
-                            this.queryForwardMap[original] = [executed, remaining];
-                        }
-                    }
+            async IAsyncEnumerable<IQuery> ExecuteAgainstCacheAsync() {
+                var cacheExecutionResult = await this.cacheExecutor.ExecuteAsync(queriesStillToExecute, cancellationToken);
+                foreach (var executedQuery in cacheExecutionResult.ExecutedQueries) {
+                    this.cacheExecutorQueries.Add(executedQuery);
+                    yield return executedQuery;
+                }
 
-                    queriesStillToExecute = cacheExecutionResult.NonExecutedQueries.Union(cacheExecutionResult.PartiallyExecutedQueries.Select(pq => pq.Remaining)).ToList();
-                    if (!queriesStillToExecute.Any()) {
-                        break;
-                    }
+                foreach (var (original, executed, remaining) in cacheExecutionResult.PartiallyExecutedQueries) {
+                    this.queryForwardMap[original] = [executed, remaining];
                 }
+
+                queriesStillToExecute = cacheExecutionResult.NonExecutedQueries.Union(cacheExecutionResult.PartiallyExecutedQueries.Select(pq => pq.Remaining)).ToList();
             }
 
             if (queriesStillToExecute.Any()) {
                 if (this.persistenceQueryExecutor == null) {
                     throw new Exception("No persistence query mechanism has been configured");
                 }
 
+                if (this.cacheExecutor != null) {
+                    // stampede protection, we take a lock on all cacheable queries
+                    var cacheableQueries = queriesStillToExecute.Where(q => q.CacheKey != null).ToArray();
+                    var locks = new Dictionary<string, IDisposable>();
+                    var locker = new AsyncDuplicateLock();
+                    try {
+                        foreach (var cacheableQuery in cacheableQueries) {
+                            if (!locks.ContainsKey(cacheableQuery.CacheKey)) {
+                                locks.Add(cacheableQuery.CacheKey, await locker.LockAsync(cacheableQuery.CacheKey));
```

Comment:
Will it make any speed difference that we are locking all queries before executing any of them?